### DISCUSSION TITLE
When running in worker mode, OpenTelemetry hook calls don't seem to work

### DISCUSSION THREAD

**Discussion by taisph** [php/frankenphp#920] [Q&A]
I don't know if I've overlooked something but using packages like `open-telemetry/opentelemetry-auto-symfony` with the opentelemetry PHP extension don't seem to work when running FrankenPHP in worker mode. As soon as I disable worker mode, the auto instrumentation starts to work again.

Anyone else stumbled upon something similar?

**Comment by dunglas** [comment]
This extension will likely have to add custom code to support the worker mode. Technically, in worker mode, a single long-running script handled many requests, so the extension code will likely be called when the worker stops, and not after each request.

**Reply by taisph** [reply]
I get the shutting down part but some of the hooks has code that is supposed to be called both before and after a Symfony kernel event, intercepting a request and handling incoming HTTP headers before a Symfony controller gets it, but that doesn't seem to happen either. It is almost as if the code adding the hooks isn't run at all. It relies on composer auto discovery but presumably this should work the same in worker mode.

**Reply by withinboredom** [reply]
Worker mode never calls any registered shutdown functions (because it doesn't shutdown) so if the extension/code/whatever registers a shutdown function ... [it won't be called](https://github.com/open-telemetry/opentelemetry-php/blob/17ad56a3f3a64481b6764bb46b5584d7bbc5b138/src/SDK/Common/Util/ShutdownHandler.php#L50). Same with the symfony kernel shutdown event (`kernel.terminate`, IIRC).

Looking at their code, it uses global static state to function ... and in worker mode, global static state is shared among all requests on that worker. So, they need to refactor the code to not use global static state and avoid using the shutdown function.

**Reply by dunglas** [reply]
> Same with the symfony kernel shutdown event (kernel.terminate, IIRC).

The Symfony kernel `terminate` event is called because there is special code to trigger it in the Symfony Runtime component: https://github.com/php-runtime/frankenphp-symfony/blob/main/src/Runner.php#L48

**Reply by Rainrider** [reply]
It works for me in worker mode using the simple span processor. The default batch span processor does not seem to work, I don't know why. See https://github.com/Rainrider/otel-symfony-frankenphp for a simple example (you need to `composer install` locally and then `docker compose up -d`).

**Reply by Rainrider** [reply]
Btw the autoinstrumentation for symfony does not use symfony's kernel events but [`hook`](https://github.com/open-telemetry/opentelemetry-php-instrumentation/blob/cff5b651bcbf12b78621dc0fa40881d68a364900/ext/opentelemetry.c#L91-L106) `HttpKernel::handle` instead

**Reply by Rainrider** [reply]
@taisph actually the batch processor works as well, it just seems to be not working because the default batch size is 512 spans and you need to reach this value for it to flush. In non-worker mode the span processor is flushed when the script finishes (follow https://github.com/open-telemetry/opentelemetry-php/blob/17ad56a3f3a64481b6764bb46b5584d7bbc5b138/src/SDK/SdkAutoloader.php#L54).

You could control the batch size with the `OTEL_BSP_MAX_EXPORT_BATCH_SIZE` environment variable. Beware that this value is per worker, so assuming a round-robin worker priority, you will need `OTEL_BSP_MAX_EXPORT_BATCH_SIZE` * `numWorkers` / `spans/worker` spans to get it to flush to your collector.

I only have followed traces here but I assume the same applies for logs and metrics as well.
